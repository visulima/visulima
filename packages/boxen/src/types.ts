interface BaseOptions {
    borderColor?: (border: string, position: BorderPosition, length: number) => string;
    float?: "center" | "left" | "right";
    footerText?: string;
    footerTextColor?: (text: string) => string;
    fullscreen?: boolean | ((width: number, height: number) => { columns: number; rows: number });
    headerText?: string;
    headerTextColor?: (text: string) => string;
    height?: number;
    textColor?: (text: string) => string;
    transformTabToSpace?: number | false;
    width?: number;
}

export type Spacer = {
    bottom: number;
    left: number;
    right: number;
    top: number;
};

export interface BorderStyle {
    bottom?: string;
    bottomLeft?: string;
    bottomRight?: string;
    horizontal?: string;
    left?: string;
    right?: string;
    top?: string;
    topLeft?: string;
    topRight?: string;
    vertical?: string;
}

export type BorderPosition = "bottom" | "bottomLeft" | "bottomRight" | "horizontal" | "left" | "right" | "top" | "topLeft" | "topRight";

export interface Options extends BaseOptions {
    borderStyle?: BorderStyle | "arrow" | "bold" | "classic" | "double" | "doubleSingle" | "none" | "round" | "single" | "singleDouble";
    footerAlignment?: "center" | "left" | "right";
    headerAlignment?: "center" | "left" | "right";
    margin?: Partial<Spacer> | number;
    padding?: Partial<Spacer> | number;
    textAlignment?: "center" | "left" | "right";
}

export interface DimensionOptions extends BaseOptions {
    borderStyle: BorderStyle | string;
    footerAlignment: "center" | "left" | "right";
    headerAlignment: "center" | "left" | "right";
    margin: Spacer;
    padding: Spacer;
    textAlignment: "center" | "left" | "right";
}
