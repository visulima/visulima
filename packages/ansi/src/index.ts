export { default as slice } from "./slice";
