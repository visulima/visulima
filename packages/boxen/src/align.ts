/**
 * This file is ported from https://github.com/nexdrew/ansi-align into typescript.
 *
 * Copyright (c) 2016, Contributors
 */
// eslint-disable-next-line import/no-extraneous-dependencies
import stringWidth from "string-width";

const halfDiff = (maxWidth: number, currentWidth: number): number => Math.floor((maxWidth - currentWidth) / 2);

const fullDiff = (maxWidth: number, currentWidth: number): number => maxWidth - currentWidth;

// eslint-disable-next-line import/prefer-default-export
export const alignAnsi = (
    text: string[] | string,
    options: {
        align?: "center" | "left" | "right";
        pad?: string;
        split?: string;
    } = {},
): string[] | string => {
    const align = options.align ?? "center";

    // short-circuit `align: 'left'` as no-op
    if (align === "left") {
        return text;
    }

    const split = options.split ?? "\n";
    const pad = options.pad ?? " ";
    const widthDiffFunction = align === "right" ? fullDiff : halfDiff;

    let returnString = false;

    if (!Array.isArray(text)) {
        returnString = true;

        // eslint-disable-next-line no-param-reassign
        text = String(text).split(split);
    }

    let width;
    let maxWidth = 0;

    // eslint-disable-next-line no-param-reassign
    text = text
        .map((string_) => {
            // eslint-disable-next-line no-param-reassign
            string_ = String(string_);
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call
            width = stringWidth(string_);

            maxWidth = Math.max(width as number, maxWidth);

            return {
                str: string_,
                width: width as number,
            };
        })
        .map((object) => Array.from({ length: widthDiffFunction(maxWidth, object.width) + 1 }).join(pad) + object.str);

    return returnString ? text.join(split) : text;
};
