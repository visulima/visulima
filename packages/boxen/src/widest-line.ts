
import stringWidth from "string-width";

// eslint-disable-next-line import/prefer-default-export
export const widestLine = (string: string): number => {
    let lineWidth = 0;

    // eslint-disable-next-line no-restricted-syntax
    for (const line of string.split("\n")) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        lineWidth = Math.max(lineWidth, stringWidth(line) as number);
    }

    return lineWidth;
};
