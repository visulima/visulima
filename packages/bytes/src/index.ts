import { Buffer } from "node:buffer";

/**
 * Converts a Node.js Buffer to a Uint8Array.
 * Note: While Buffer instances are Uint8Array instances, this function
 * creates a new Uint8Array view on the Buffer's underlying ArrayBuffer,
 * ensuring correct byteOffset and length.
 * @param buf The Buffer to convert.
 * @returns A Uint8Array view of the Buffer.
 */
export const bufferToUint8Array = (buf: Buffer): Uint8Array => new Uint8Array(buf.buffer, buf.byteOffset, buf.length);

/**
 * Checks if a value is a Uint8Array or, in a Node.js environment, a Buffer.
 * @param x The value to check.
 * @returns True if x is a Uint8Array or Buffer, false otherwise.
 */
export const isUint8Array
  = typeof Buffer === "function"
      ? (x: unknown): x is Uint8Array => x instanceof Uint8Array || Buffer.isBuffer(x)
      : (x: unknown): x is Uint8Array => x instanceof Uint8Array;

/**
 * Converts an ASCII string, array of strings, or template literal to a Uint8Array.
 * Each character is converted to its ASCII byte value (0-255). Characters outside this range will be truncated.
 * @param txt The input string, array of strings, or template strings array.
 * @returns A Uint8Array representing the ASCII encoded string.
 */
export const asciiToUint8Array = (txt: TemplateStringsArray | string | [string]): Uint8Array => {
    if (typeof txt === "string")
        return asciiToUint8Array([txt]);

    const [input] = Array.isArray(txt) ? txt : [String.raw(txt as TemplateStringsArray)]; // Handle TemplateStringsArray
    const inputLength = input.length;
    const result = new Uint8Array(inputLength); // Renamed 'res' to 'result'

    for (let index = 0; index < inputLength; index += 1) { // Changed index++ to index += 1
        // eslint-disable-next-line unicorn/prefer-code-point, no-bitwise
        result[index] = input.charCodeAt(index) & 0xFF; // Ensure ASCII range
    }

    return result;
};

/**
 * Converts a UTF-8 string, array of strings, or template literal to a Uint8Array.
 * Requires Node.js Buffer support.
 * @param txt The input string, array of strings, or template strings array.
 * @returns A Uint8Array representing the UTF-8 encoded string.
 */
export const utf8ToUint8Array = (txt: TemplateStringsArray | [string] | string): Uint8Array => {
    if (typeof txt === "string")
        return utf8ToUint8Array([txt]);

    const [input] = Array.isArray(txt) ? txt : [String.raw(txt as TemplateStringsArray)]; // Handle TemplateStringsArray

    return bufferToUint8Array(Buffer.from(input, "utf8"));
};

/**
 * Attempts to convert various data types to a Uint8Array.
 * Supports Uint8Array (returns as is), ArrayBuffer, Array of numbers.
 * In Node.js, also supports Buffer (returns as is or converts from other types via Buffer.from).
 * @param data The data to convert.
 * @returns A Uint8Array representation of the input data.
 * @throws Error 'UINT8ARRAY_INCOMPATIBLE' if the data cannot be converted and Buffer is not available for fallback.
 */
export const toUint8Array = (data: unknown): Uint8Array => {
    if (typeof Buffer === "function" && Buffer.isBuffer(data)) {
        return bufferToUint8Array(data); // Prioritize Buffer and ensure plain Uint8Array
    }

    if (data instanceof Uint8Array) { // This will now handle non-Buffer Uint8Arrays
        return data;
    }

    if (data instanceof ArrayBuffer)
        return new Uint8Array(data);

    // Ensure it's an array of numbers before creating Uint8Array from it
    if (Array.isArray(data) && data.every((item) => typeof item === "number")) {
        return new Uint8Array(data);
    }

    if (typeof Buffer === "function") {
        if (Buffer.isBuffer(data)) {
            return bufferToUint8Array(data); // Use our converter for consistency
        }

        // Attempt to convert string via Buffer.from.
        // Other Buffer-compatible types like ArrayBuffer or ArrayBufferView (which Buffer is)
        // are typically handled by earlier checks or are Buffers themselves.
        if (typeof data === "string") {
            try {
                return bufferToUint8Array(Buffer.from(data));
            } catch {
                // If Buffer.from fails, fall through to error
            }
        }
    }

    throw new Error("UINT8ARRAY_INCOMPATIBLE: Cannot convert data to Uint8Array");
};

// eslint-disable-next-line import/no-extraneous-dependencies
export * from "@std/bytes";
