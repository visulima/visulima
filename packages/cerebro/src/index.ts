// eslint-disable-next-line @typescript-eslint/no-namespace,@typescript-eslint/no-unused-vars
declare namespace NodeJS {
    interface ProcessEnvironment {
        CEREBRO_MIN_NODE_VERSION?: string;
        CEREBRO_OUTPUT_LEVEL: VERBOSITY_LEVEL;
    }
}

export type { Cli } from "./@types/cli";
export type { ArgumentDefinition, Command, OptionDefinition } from "./@types/command";
export type { Extension } from "./@types/extension";
export type { Toolbox } from "./@types/toolbox";

/**
 * Any of the output types [[OUTPUT_NORMAL]], [[OUTPUT_RAW]] and [[OUTPUT_PLAIN]].
 */
export type OutputType = 1 | 2 | 4;

/**
 * Any of the verbosity types
 * [[VERBOSITY_QUIET]], [[VERBOSITY_NORMAL]], [[VERBOSITY_VERBOSE]] and [[VERBOSITY_DEBUG]].
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export type VERBOSITY_LEVEL = 16 | 32 | 64 | 128 | 256;

export { default } from "./cli";
