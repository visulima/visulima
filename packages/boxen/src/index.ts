/**
 * This file is a modified version of the original `boxen` package.
 *
 * MIT License
 *
 * Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (https://sindresorhus.com)
 */

import cliBoxes from "cli-boxes";
import stringWidth from "string-width";
import terminalSize from "terminal-size";
import wrapAnsi from "wrap-ansi";

import { alignAnsi } from "./align";
import type { BorderPosition, BorderStyle, DimensionOptions, Options, Spacer } from "./types";
import { widestLine } from "./widest-line";

const NEWLINE = "\n";
const PAD = " ";
const NONE = "none";

const getObject = (detail: Partial<Spacer> | number | undefined): Spacer =>
    (typeof detail === "number"
        ? {
              bottom: detail,
              left: detail * 3,
              right: detail * 3,
              top: detail,
          }
        : {
              bottom: 0,
              left: 0,
              right: 0,
              top: 0,
              ...detail,
          });

const getBorderWidth = (borderStyle: BorderStyle | string) => (borderStyle === NONE ? 0 : 2);

// eslint-disable-next-line sonarjs/cognitive-complexity
const getBorderChars = (borderStyle: BorderStyle | string): BorderStyle => {
    const sides = ["topLeft", "topRight", "bottomRight", "bottomLeft", "left", "right", "top", "bottom"];

    let characters: BorderStyle;

    // Create empty border style
    if (borderStyle === NONE) {
        // eslint-disable-next-line no-param-reassign
        borderStyle = {};

        // eslint-disable-next-line no-restricted-syntax
        for (const side of sides) {
            // eslint-disable-next-line no-param-reassign
            borderStyle[side as keyof BorderStyle] = "";
        }
    }

    if (typeof borderStyle === "string") {
        const cliBox = cliBoxes[borderStyle as keyof typeof cliBoxes] as BorderStyle | undefined;

        if (cliBox === undefined) {
            throw new TypeError(`Invalid border style: ${borderStyle}`);
        }

        characters = cliBox;
    } else {
        // Ensure retro-compatibility
        if (typeof borderStyle.vertical === "string") {
            // eslint-disable-next-line no-param-reassign
            borderStyle.left = borderStyle.vertical;
            // eslint-disable-next-line no-param-reassign
            borderStyle.right = borderStyle.vertical;
        }

        // Ensure retro-compatibility
        if (typeof borderStyle.horizontal === "string") {
            // eslint-disable-next-line no-param-reassign
            borderStyle.top = borderStyle.horizontal;
            // eslint-disable-next-line no-param-reassign
            borderStyle.bottom = borderStyle.horizontal;
        }

        // eslint-disable-next-line no-restricted-syntax
        for (const side of sides) {
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            if (borderStyle[side as keyof BorderStyle] === null || typeof borderStyle[side as keyof BorderStyle] !== "string") {
                throw new TypeError(`Invalid border style: ${side}`);
            }
        }

        characters = borderStyle;
    }

    return characters;
};

const wrapText = (
    text: string,
    colorizeText: (string: string) => string,
    horizontal: string,
    colorizeBorder: (string: string, length: number) => string,
    alignment: string,
) => {
    let title = "";

    // eslint-disable-next-line no-param-reassign
    text = colorizeText(text);

    const textWidth = stringWidth(text);

    switch (alignment) {
        case "left": {
            title = text + colorizeBorder(horizontal.slice(textWidth), stringWidth(horizontal.slice(textWidth)));

            break;
        }

        case "right": {
            title = colorizeBorder(horizontal.slice(textWidth + 2), stringWidth(horizontal.slice(textWidth)) + 2) + " " + text + " ";

            break;
        }

        default: {
            // eslint-disable-next-line no-param-reassign
            horizontal = horizontal.slice(textWidth);

            if (horizontal.length % 2 === 1) {
                // This is needed in case the length is odd
                // eslint-disable-next-line no-param-reassign
                horizontal = horizontal.slice(Math.floor(horizontal.length / 2));

                title = colorizeBorder(horizontal.slice(1), stringWidth(horizontal.slice(1))) + text + colorizeBorder(horizontal, stringWidth(horizontal)); // We reduce the left part of one character to avoid the bar to go beyond its limit
            } else {
                // eslint-disable-next-line no-param-reassign
                horizontal = horizontal.slice(horizontal.length / 2);

                const horizontalLength = stringWidth(horizontal);

                title = colorizeBorder(horizontal, horizontalLength) + text + colorizeBorder(horizontal, horizontalLength);
            }

            break;
        }
    }

    return title;
};

type ContentTextOptions = {
    height: number | undefined;
    padding: Spacer;
    textAlignment: "center" | "left" | "right";
    width: number;
};

const makeContentText = (
    text: string,
    { height, padding, textAlignment, width }: ContentTextOptions,
    // eslint-disable-next-line sonarjs/cognitive-complexity
) => {
    // eslint-disable-next-line no-param-reassign
    text = alignAnsi(text, { align: textAlignment }) as string;

    let lines: string[] = text.split(NEWLINE);

    const textWidth = widestLine(text);
    const max = width - padding.left - padding.right;

    if (textWidth > max) {
        const newLines = [];

        // eslint-disable-next-line no-restricted-syntax
        for (const line of lines) {
            const createdLines = wrapAnsi(line, max, { hard: true }) as string;
            const alignedLines = alignAnsi(createdLines, { align: textAlignment });
            const alignedLinesArray = (alignedLines as string).split("\n");
            const longestLength = Math.max(...alignedLinesArray.map((s) => stringWidth(s) as number));

            // eslint-disable-next-line no-restricted-syntax
            for (const alignedLine of alignedLinesArray) {
                let paddedLine;

                switch (textAlignment) {
                    case "center": {
                        paddedLine = PAD.repeat((max - longestLength) / 2) + alignedLine;
                        break;
                    }

                    case "right": {
                        paddedLine = PAD.repeat(max - longestLength) + alignedLine;
                        break;
                    }

                    default: {
                        paddedLine = alignedLine;
                        break;
                    }
                }

                newLines.push(paddedLine);
            }
        }

        lines = newLines;
    }

    if (textAlignment === "center" && textWidth < max) {
        lines = lines.map((line) => PAD.repeat((max - textWidth) / 2) + line);
    } else if (textAlignment === "right" && textWidth < max) {
        lines = lines.map((line) => PAD.repeat(max - textWidth) + line);
    }

    const paddingLeft = PAD.repeat(padding.left);
    const paddingRight = PAD.repeat(padding.right);

    lines = lines.map((line) => {
        const newLine = paddingLeft + line + paddingRight;

        return newLine + PAD.repeat(width - stringWidth(newLine));
    });

    if (padding.top > 0) {
        lines = [...Array.from<string>({ length: padding.top }).fill(PAD.repeat(width)), ...lines];
    }

    if (padding.bottom > 0) {
        lines = [...lines, ...Array.from<string>({ length: padding.bottom }).fill(PAD.repeat(width))];
    }

    if (height && lines.length > height) {
        lines = lines.slice(0, height);
    } else if (height && lines.length < height) {
        lines = [...lines, ...Array.from<string>({ length: height - lines.length }).fill(PAD.repeat(width))];
    }

    return lines.join(NEWLINE);
};

// eslint-disable-next-line sonarjs/cognitive-complexity
const boxContent = (content: string, contentWidth: number, columnsWidth: number, options: DimensionOptions): string => {
    const colorizeBorder = (border: string, position: BorderPosition, length: number): string =>
        (options.borderColor ? options.borderColor(border, position, length) : border);
    const colorizeHeaderText = (title: string): string => (options.headerTextColor ? options.headerTextColor(title) : title);
    const colorizeFooterText = (title: string): string => (options.footerTextColor ? options.footerTextColor(title) : title);
    const colorizeContent = (value: string): string => (options.textColor ? options.textColor(value) : value);

    const chars = getBorderChars(options.borderStyle) as Required<BorderStyle>;

    let marginLeft = PAD.repeat(options.margin.left);

    if (options.float === "center") {
        const marginWidth = Math.max((columnsWidth - contentWidth - getBorderWidth(options.borderStyle)) / 2, 0);
        marginLeft = PAD.repeat(marginWidth);
    } else if (options.float === "right") {
        const marginWidth = Math.max(columnsWidth - contentWidth - options.margin.right - getBorderWidth(options.borderStyle), 0);
        marginLeft = PAD.repeat(marginWidth);
    }

    let result = "";

    if (options.margin.top) {
        result += NEWLINE.repeat(options.margin.top);
    }

    if (options.borderStyle !== NONE || options.headerText) {
        let headerText = colorizeBorder(chars.top.repeat(contentWidth), "top", contentWidth);

        if (options.headerText) {
            headerText = wrapText(
                options.headerText,
                colorizeHeaderText,
                chars.top.repeat(contentWidth),
                (value: string, length: number) => colorizeBorder(value, "top", length),
                options.headerAlignment,
            );
        }

        const topBorder = colorizeBorder(marginLeft + chars.topLeft, "topLeft", stringWidth(marginLeft + chars.topLeft));

        result += topBorder + headerText + colorizeBorder(chars.topRight, "topRight", stringWidth(chars.topRight)) + NEWLINE;
    }

    const lines = content.split(NEWLINE);

    result += lines
        .map(
            (line) =>
                marginLeft +
                colorizeBorder(chars.left, "left", stringWidth(chars.left)) +
                colorizeContent(line) +
                colorizeBorder(chars.right, "right", stringWidth(chars.right)),
        )
        .join(NEWLINE);

    if (options.borderStyle !== NONE || options.footerText) {
        const bottomBorder = NEWLINE + colorizeBorder(marginLeft + chars.bottomLeft, "bottomLeft", stringWidth(marginLeft + chars.bottomLeft));
        let footerText = colorizeBorder(chars.bottom.repeat(contentWidth), "bottom", contentWidth);

        if (options.footerText) {
            footerText = wrapText(
                options.footerText,
                colorizeFooterText,
                chars.bottom.repeat(contentWidth),
                (value: string, length: number) => colorizeBorder(value, "bottom", length),
                options.footerAlignment,
            );
        }

        result += bottomBorder + footerText + colorizeBorder(chars.bottomRight, "bottomRight", stringWidth(chars.bottomRight));
    }

    if (options.margin.bottom) {
        result += NEWLINE.repeat(options.margin.bottom);
    }

    return result;
};

const sanitizeOptions = (options: DimensionOptions): DimensionOptions => {
    // If fullscreen is enabled, max-out unspecified width/height
    if (options.fullscreen) {
        let newDimensions = terminalSize();

        if (typeof options.fullscreen === "function") {
            newDimensions = options.fullscreen(newDimensions.columns, newDimensions.rows);
        }

        if (!options.width) {
            // eslint-disable-next-line no-param-reassign
            options.width = newDimensions.columns;
        }

        if (!options.height) {
            // eslint-disable-next-line no-param-reassign
            options.height = newDimensions.rows;
        }
    }

    // If width is provided, make sure it's not below 1
    if (options.width) {
        // eslint-disable-next-line no-param-reassign
        options.width = Math.max(1, options.width - getBorderWidth(options.borderStyle));
    }

    // If height is provided, make sure it's not below 1
    if (options.height) {
        // eslint-disable-next-line no-param-reassign
        options.height = Math.max(1, options.height - getBorderWidth(options.borderStyle));
    }

    return options;
};

const formatTitle = (title: string, borderStyle: BorderStyle | string): string => (borderStyle === NONE ? title : ` ${title} `);

// eslint-disable-next-line sonarjs/cognitive-complexity
const determineDimensions = (text: string, columnsWidth: number, options: DimensionOptions): DimensionOptions => {
    // eslint-disable-next-line no-param-reassign
    options = sanitizeOptions(options);

    const widthOverride = options.width !== undefined;
    const borderWidth = getBorderWidth(options.borderStyle);
    const maxWidth = columnsWidth - options.margin.left - options.margin.right - borderWidth;

    const widest = widestLine(wrapAnsi(text, columnsWidth - borderWidth, { hard: true, trim: false })) + options.padding.left + options.padding.right;

    // If title and width are provided, title adheres to fixed width
    if (options.headerText && widthOverride) {
        // eslint-disable-next-line no-param-reassign
        options.headerText = options.headerText.slice(0, Math.max(0, (options.width as number) - 2));

        if (options.headerText) {
            // eslint-disable-next-line no-param-reassign
            options.headerText = formatTitle(options.headerText, options.borderStyle);
        }
    } else if (options.headerText) {
        // eslint-disable-next-line no-param-reassign
        options.headerText = options.headerText.slice(0, Math.max(0, maxWidth - 2));

        // Recheck if title isn't empty now
        if (options.headerText) {
            // eslint-disable-next-line no-param-reassign
            options.headerText = formatTitle(options.headerText, options.borderStyle);
            // If the title is larger than content, box adheres to title width
            if (stringWidth(options.headerText) > widest) {
                // eslint-disable-next-line no-param-reassign
                options.width = stringWidth(options.headerText);
            }
        }
    }

    // If fixed width is provided, use it or content width as reference
    // eslint-disable-next-line no-param-reassign,@typescript-eslint/prefer-nullish-coalescing
    options.width = options.width || widest;

    if (!widthOverride) {
        if (options.margin.left && options.margin.right && options.width > maxWidth) {
            // Let's assume we have margins: left = 3, right = 5, in total = 8
            const spaceForMargins = columnsWidth - options.width - borderWidth;
            // Let's assume we have space = 4
            const multiplier = spaceForMargins / (options.margin.left + options.margin.right);
            // Here: multiplier = 4/8 = 0.5
            // eslint-disable-next-line no-param-reassign
            options.margin.left = Math.max(0, Math.floor(options.margin.left * multiplier));
            // eslint-disable-next-line no-param-reassign
            options.margin.right = Math.max(0, Math.floor(options.margin.right * multiplier));
            // Left: 3 * 0.5 = 1.5 -> 1
            // Right: 6 * 0.5 = 3
        }

        // Re-cap width considering the margins after shrinking
        // eslint-disable-next-line no-param-reassign
        options.width = Math.min(options.width, columnsWidth - borderWidth - options.margin.left - options.margin.right);
    }

    // Prevent padding overflow
    if (options.width - (options.padding.left + options.padding.right) <= 0) {
        // eslint-disable-next-line no-param-reassign
        options.padding.left = 0;
        // eslint-disable-next-line no-param-reassign
        options.padding.right = 0;
    }

    if (options.height && options.height - (options.padding.top + options.padding.bottom) <= 0) {
        // eslint-disable-next-line no-param-reassign
        options.padding.top = 0;
        // eslint-disable-next-line no-param-reassign
        options.padding.bottom = 0;
    }

    return options;
};

// eslint-disable-next-line import/prefer-default-export
export const boxen = (text: string, options: Options = {}): string => {
    if (options.borderColor !== undefined && typeof options.borderColor !== "function") {
        throw new Error(`"borderColor" is not a valid function`);
    }

    if (options.textColor !== undefined && typeof options.textColor !== "function") {
        throw new Error(`"backgroundColor" is not a valid function`);
    }

    let config: DimensionOptions = {
        borderStyle: "single",
        dimBorder: false,
        float: "left",
        footerAlignment: "right",
        headerAlignment: "left",
        textAlignment: "left",
        transformTabToSpace: 4,
        ...options,
    } as DimensionOptions;

    config.padding = getObject(options.padding ?? 0);
    config.margin = getObject(options.margin);

    // replace tabs with spaces
    if (config.transformTabToSpace) {
        // eslint-disable-next-line no-param-reassign
        text = text.replaceAll("\t", " ".repeat(config.transformTabToSpace));
    }

    const { columns } = terminalSize();

    config = determineDimensions(text, columns, config);

    return boxContent(makeContentText(text, config as ContentTextOptions), config.width as number, columns, config);
};
