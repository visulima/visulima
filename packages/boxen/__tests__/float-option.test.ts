import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { boxen } from "../src";

vi.mock("terminal-size", () => {
    return {
        default: () => {
            return {
                columns: 80,
                rows: 24,
            };
        },
    };
});

describe("float option", () => {
    beforeEach(() => {
        vi.stubGlobal("process", {
            env: { COLUMNS: "22" },
            stderr: {
                rows: 24,
            },
            stdout: {
                columns: 80,
            },
        });
    });

    afterEach(() => {
        vi.unstubAllGlobals();
    });

    it("float option (left)", () => {
        expect.assertions(1);

        const box = boxen("foo", {
            float: "left",
        });

        expect(box).toMatchSnapshot();
    });

    it("float option (center)", () => {
        expect.assertions(1);

        const box = boxen("foo", {
            float: "center",
        });

        expect(box).toMatchSnapshot();
    });

    it("float option (right)", () => {
        expect.assertions(1);

        const box = boxen("foo", {
            float: "right",
        });

        expect(box).toMatchSnapshot();
    });

    it("float option (center) with margin", () => {
        expect.assertions(1);

        const box = boxen("foo", {
            float: "right",
            margin: {
                left: 3,
                top: 4,
            },
        });

        expect(box).toMatchSnapshot();
    });

    it("float option (right) with margin", () => {
        expect.assertions(1);

        const box = boxen("foo", {
            float: "right",
            margin: {
                bottom: 5,
                right: 2,
            },
        });

        expect(box).toMatchSnapshot();
    });

    it("float option (center) when content > columns", () => {
        expect.assertions(2);

        const longContent = "foobar".repeat(22);

        expect(() => {
            boxen(longContent, {
                float: "center",
            });
        }).not.toThrow();

        const box = boxen(longContent, {
            float: "center",
        });

        expect(box).toMatchSnapshot();
    });

    it("float option (right) when content > columns", () => {
        expect.assertions(2);

        const longContent = "foobar".repeat(22);

        expect(() => {
            boxen(longContent, {
                float: "right",
            });
        }).not.toThrow();

        const box = boxen(longContent, {
            float: "right",
        });

        expect(box).toMatchSnapshot();
    });
});
