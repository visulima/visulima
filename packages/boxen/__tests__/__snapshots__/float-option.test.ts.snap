// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`float option > float option (center) 1`] = `
"                                     ┌───┐
                                     │foo│
                                     └───┘"
`;

exports[`float option > float option (center) when content > columns 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────┐
│foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar│
│foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar                        │
└──────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`float option > float option (center) with margin 1`] = `
"



                                                                           ┌───┐
                                                                           │foo│
                                                                           └───┘"
`;

exports[`float option > float option (left) 1`] = `
"┌───┐
│foo│
└───┘"
`;

exports[`float option > float option (right) 1`] = `
"                                                                           ┌───┐
                                                                           │foo│
                                                                           └───┘"
`;

exports[`float option > float option (right) when content > columns 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────┐
│foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar│
│foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar                        │
└──────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`float option > float option (right) with margin 1`] = `
"                                                                         ┌───┐
                                                                         │foo│
                                                                         └───┘




"
`;
